ospfd/ospf_packet.c

```diff
@@ -2947,225 +2942,250 @@ int ospf_read(struct thread *thread)
 	uint16_t length;
 	struct interface *ifp = NULL;
 	struct connected *c;
+	int32_t count = 0;
 
 	/* first of all get interface pointer. */
 	ospf = THREAD_ARG(thread);
 
 	/* prepare for next packet. */
-	ospf->t_read = NULL;
 	thread_add_read(master, ospf_read, ospf, ospf->fd, &ospf->t_read);
 
-	stream_reset(ospf->ibuf);
-	ibuf = ospf_recv_packet(ospf, ospf->fd, &ifp, ospf->ibuf);
-	if (ibuf == NULL)
-		return -1;
-	/* This raw packet is known to be at least as big as its IP header. */
+	while (count < ospf->write_oi_count) {
+		stream_reset(ospf->ibuf);
+		ibuf = ospf_recv_packet(ospf, ospf->fd, &ifp, ospf->ibuf);
+		if (ibuf == NULL)
+			return -1;
+		/*
+		 * This raw packet is known to be at least as big as its
+		 * IP header.
+		 * Note that there should not be alignment problems with
+		 * this assignment because this is at the beginning of the
+		 * stream data buffer.
+		 */
+		iph = (struct ip *)STREAM_DATA(ibuf);
+		/* Note that sockopt_iphdrincl_swab_systoh was called in
+		 * ospf_recv_packet. */
+
+		if (ifp == NULL) {
+			/* Handle cases where the platform does not support
+			 * retrieving the ifindex, and also platforms (such
+			 * as Solaris 8) that claim to support ifindex
+			 * retrieval but do not.
+			 */
+			c = if_lookup_address((void *)&iph->ip_src, AF_INET,
+					      ospf->vrf_id);
+			if (c)
+				ifp = c->ifp;
+			if (ifp == NULL) {
+				if (IS_DEBUG_OSPF_PACKET(0, RECV))
+					zlog_debug(
+						"%s: Unable to determine incoming interface from: %s(%s)",
+						__PRETTY_FUNCTION__,
+						inet_ntoa(iph->ip_src),
+						ospf->name ? ospf->name
+							   : VRF_DEFAULT_NAME);
+				continue;
+			}
+		}
 
-	/* Note that there should not be alignment problems with this assignment
-	   because this is at the beginning of the stream data buffer. */
-	iph = (struct ip *)STREAM_DATA(ibuf);
-	/* Note that sockopt_iphdrincl_swab_systoh was called in
-	 * ospf_recv_packet. */
-
-	if (ifp == NULL) {
-		/* Handle cases where the platform does not support retrieving
-		   the ifindex,
-		   and also platforms (such as Solaris 8) that claim to support
-		   ifindex
-		   retrieval but do not. */
-		c = if_lookup_address((void *)&iph->ip_src, AF_INET,
-				      ospf->vrf_id);
-		if (c)
-			ifp = c->ifp;
-		if (ifp == NULL)
-			return 0;
-	}
+		/* Self-originated packet should be discarded silently. */
+		if (ospf_if_lookup_by_local_addr(ospf, NULL, iph->ip_src)) {
+			if (IS_DEBUG_OSPF_PACKET(0, RECV)) {
+				zlog_debug(
+					"ospf_read[%s]: Dropping self-originated packet",
+					inet_ntoa(iph->ip_src));
+			}
+			continue;
```

Comment:
All the place where we continue we may have to increment count and also move buffer ahead?

Reply:
already answered but moving ibuf ahead is not needed